### ISSUE TITLE
updating cri-o from 1.29.9 to 1.30.6

### ISSUE THREAD

**Issue by ivanweb568** [cri-o/cri-o#8705]
### What happened?

With the cri-o version(in 1.29), I see the default runtime is set to crun in /etc/crio/crio.conf.d/10-crio.conf
I had updated the cri-o version from 1.28.9 to 1.29.9 successfully with runc runtime, but while upgrading cri-o from 1.29.9 to 1.30.6, I tried the same to keep the default runtime to runc and once when the new crio rpm cri-o-1.30.6 was installed few containers came up with crun runtime, I updated the crio config file to change runtime to runc and restarted crio  but I can still see some container processes with crun on K8S worker node.

some k8s pods are throwing below "address already in use" error, it is due to the crun process still running and not being able to terminate the container process
below reference is for calico pod
 felix/health.go 392: Health endpoint failed, trying to restart it... error=listen tcp 127.0.0.1:9099: bind: address already in use

I was hoping to keep runtime as cri-o in this release and move to crun to next 1.31 release as crun will be default runtime for further releases as well

I also tried to tried to migrate to crun, but some runc container processes are still up causing above issue

### What did you expect to happen?

when making change from runc to crio, it should not affect kubernetes pods status

### How can we reproduce it (as minimally and precisely as possible)?

on k8s node using cri-o, change runtime from runc to crun in cri-o-1.30.6 version and restart crio

### CRI-O and Kubernetes version

<details>

```console
$ crio --version
crio version 1.30.6
Version:        1.30.6
GitCommit:      747ac86f851b1d0b3ce2cc09a4b7c053c6ead8ff
GitCommitDate:  2024-10-01T19:44:25Z
GitTreeState:   clean
BuildDate:      1970-01-01T00:00:00Z
GoVersion:      go1.22.0
Compiler:       gc
Platform:       linux/amd64
Linkmode:       static
BuildTags:
  static
  netgo
  osusergo
  exclude_graphdriver_btrfs
  exclude_graphdriver_devicemapper
  seccomp
  apparmor
  selinux
LDFlags:          unknown
SeccompEnabled:   true
AppArmorEnabled:  false
```

```console
$ kubectl version --output=json
{
  "clientVersion": {
    "major": "1",
    "minor": "30",
    "gitVersion": "v1.30.5",
    "gitCommit": "74e84a90c725047b1328ff3d589fedb1cb7a120e",
    "gitTreeState": "clean",
    "buildDate": "2024-09-12T00:18:35Z",
    "goVersion": "go1.22.6",
    "compiler": "gc",
    "platform": "linux/amd64"
  },
  "kustomizeVersion": "v5.0.4-0.20230601165947-6ce0bf390ce3"
}

```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Oracle Linux Server"
VERSION="8.10"
ID="ol"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="8.10"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Oracle Linux Server 8.10"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:oracle:linux:8:10:server"
HOME_URL="https://linux.oracle.com/"
BUG_REPORT_URL="https://github.com/oracle/oracle-linux"

ORACLE_BUGZILLA_PRODUCT="Oracle Linux 8"
ORACLE_BUGZILLA_PRODUCT_VERSION=8.10
ORACLE_SUPPORT_PRODUCT="Oracle Linux"
ORACLE_SUPPORT_PRODUCT_VERSION=8.10

$ uname -a
Linux 5.4.17-2136.334.6.1.el8uek.x86_64 #3 SMP Mon Aug 12 16:49:24 PDT 2024 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Additional environment details (AWS, VirtualBox, physical, etc.)

<details>

</details>

**Comment by ivanweb568** [comment]
typo correction
What did you expect to happen?
when making change from runc to **crun**, it should not affect kubernetes pods status

**Comment by haircommander** [comment]
cri-o should have all of the available information to restore the containers and remove them with the correct runtime handler, so this is a bug, but generally before restarting cri-o and making a change like this i'd recommend draining the node and stopping the kubelet first

**Comment by dipaksawle** [comment]
@haircommander  We are facing the same issue. Drained the kubenode but draining kubenode does not delete the damonset pods like calico. Then I stopped the kubelet and crio and changed the runtime to "crun" in crio.conf but still container processes with runc are running and and new pods are not starting up due to error "error=listen tcp 127.0.0.1:9099: bind: address already in use"

**Comment by haircommander** [comment]
ugh yeah forgot about daemonsets. I think the cleanest way to to this is do a reboot before restarting. If not, then you also can manually use crictl to remove the daemonsets before restarting.

**Comment by dipaksawle** [comment]
Yes. Reboot is the cleanest way but in production we have hundreds of kubenodes and drain+reboot of each node would take hours for deployment to complete.

**Comment by dipaksawle** [comment]
@haircommander  Is there a possibility to get us a timeline to get this resolved? This is one bug that we found and other people may see a different issues, Since it was missed in the documentation to update the default runtime to CRUN in 1.30 not 1.31, there may be some other changes as well. This bug if it exists in 1.30 then it exists in 1.31 as well. That means if anyone running CRIO 1.3X.X, is running a buggy version. Which may fail for one or other reason if not fixed ASAP. Please correct me if I am wrong anywhere.

**Comment by kwilczynski** [comment]
> Thank you @kwilczynski for sharing detailed steps.

@dipaksawle, no problem. I hope these help.

> Also, Changing "CRUN" as the default runtime is mentioned in release notes of CRIO 1.31. Its not mentioned in 1.30 release notes. Looks like somehow its missed to update in 1.30.X release notes.

I think this is something we have fumbled with the update to the default configuration file, rather than CRI-O in itself, that went through as part of the binary packages changes per the following:

- https://github.com/cri-o/packaging/pull/23

This is something we should probably fix for 1.30. I apologise for the trouble this has caused.

/cc @saschagrunert

**Comment by dipaksawle** [comment]
Its not only about /etc/crio/crio.conf.d/10-crio.conf file which comes with crio RPM.
This file has default_runtime = "crun"
So, before restarting crio service, we even changed default_runtime to "runc" but still some container processes started with crun.
So, I think it needs to be fixed in code as well.

**Comment by saschagrunert** [comment]
> This is something we should probably fix for 1.30.

We would need to maintain different configuration templates for supported release branches in https://github.com/cri-o/packaging/tree/main/templates/latest

**Comment by kwilczynski** [comment]
> @kwilczynski Its not only about /etc/crio/crio.conf.d/10-crio.conf file which comes with crio RPM. This file has default_runtime = "crun" So, before restarting crio service, we even changed default_runtime to "runc" but still some container processes started with crun. So, I think it needs to be fixed in code as well.

@dipaksawle, there are two parts to it:

- Unexpected configuration file change (especially for 1.30)
- CRI-O does not handle runtime change in realtime

The latter is a known behaviour which, as I said, we will fix at some point in the future releases.
